### PR TITLE
Enable go-sockaddr templating for `network-interface`

### LINKED ISSUE #3675: client network_interface config doesn't parse sockaddr templates
### Nomad version
0.7.1

### Operating system and Environment details

Ubuntu 16.04

### Issue

Unable to specify network_interface option for an alias interface, eth0:1

### Reproduction steps

Assign an interface eth0:1 (Linode uses these for private address space) 

```
eth0:1    Link encap:Ethernet  HWaddr f2:3c:23:b1:45:ff
          inet addr:192.168.122.21  Bcast:0.0.0.0  Mask:255.255.128.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
```

```
/usr/local/bin/nomad agent -config=/etc/nomad -log-level=DEBUG -network-interface=eth0:1
    Loaded configuration from /etc/nomad/config.json
==> Starting Nomad agent...
==> Error starting agent: client setup failed: fingerprinting failed: Error while detecting network interface during fingerprinting: route ip+net: no such network interface
    2017/12/19 23:25:33.387588 [INFO] client: using state directory /opt/nomad/client
    2017/12/19 23:25:33.387751 [INFO] client: using alloc directory /opt/nomad/alloc
    2017/12/19 23:25:33.390788 [DEBUG] client: built-in fingerprints: [arch cgroup consul cpu host memory network nomad signal storage vault env_gce env_aws]
    2017/12/19 23:25:33.390971 [INFO] fingerprint.cgroups: cgroups are available
    2017/12/19 23:25:33.391097 [DEBUG] client: fingerprinting cgroup every 15s
    2017/12/19 23:25:33.392795 [INFO] fingerprint.consul: consul agent is available
    2017/12/19 23:25:33.392955 [DEBUG] client: fingerprinting consul every 15s
    2017/12/19 23:25:33.392966 [DEBUG] fingerprint.cpu: frequency: 2799 MHz
    2017/12/19 23:25:33.392970 [DEBUG] fingerprint.cpu: core count: 2
```

### LINKED ISSUE #5498: Network fingerprinting on lo
### Nomad version
`Nomad v0.9.0-rc1 (7c00ab4f3f37cfd1e258b38fd2ad99e7bc23e4c3)` and I assume everything below

### Operating system and Environment details
Linux

### Issue
In a clos network (L3, unnumbered BGP routed fabric) setup, it is common to use the loopback interface (`lo`) to assign the node address. 

This loopback IP then appears as a connected address, and can be announced into the BGP fabric by a BGP speaker on the host.

Unfortunately, when configuring the `client -> network_interface` setting in nomad to `lo`, the `unique.network.ip-address` allocated to the host is *always* `127.0.0.1` - this is because the fingerprinter [always picks the first IP address returned on the interface](https://github.com/hashicorp/nomad/blob/master/client/fingerprint/network.go#L112), which for `lo` is always going to be (and must be) the actual loopback address.

This breaks service discovery as services are announced into consul using `127.0.0.1`, which is only valid on the node currently running that service.

Am I missing some way to override the `unique.network.ip-address` setting?

It seems to me like setting `127.0.0.1` as `unique.network.ip-address` is almost always the incorrect thing to do when fingerprinting, unless running a single-node dev setup.

IMO the correct fix is to modify the fingerprinting code to prefer any other applied address on `lo` over `127.0.0.1/8`. I also have to specifically set the `bind_addr` option to the correct loopback IP, so it would seem that defaulting back to the value of `bind_addr` if set is more sane than allocating `127.0.0.1`.

I think it's probably possible to work around this by creating a dummy interface with the same IP as the loopback IP and pointing nomad at that, but would be great to fix this in nomad instead.

Happy to put a PR together for a fix if this is agreed as the correct approach?

### PR DESCRIPTION
**Pull request by angrycub**
This adds go-sockaddr templating for the client.network_interface attribute. @lgfa29 helped me write the tests.

Closes #3675
Fixes #5498

### CODE REVIEW

File: command/agent/config_test.go [hunk 1]
```diff
@@ -984,6 +985,95 @@ func TestConfig_normalizeAddrs(t *testing.T) {
...
+			RPC:  "127.0.0.1:4647",
+			Serf: "127.0.0.1:4648",
+		},
+		DevMode: false,
+		Client: &ClientConfig{
+			Enabled:          true,
+			NetworkInterface: "not an interface",
+		},
+	}
+	if err := c.normalizeAddrs(); err == nil {
```
Comment (tgross) [inline command/agent/config_test.go:1074]: We've been trying to gradually switch to use testify for tests like this. So `require.NoError(t, c.normalizeAddrs())`. This gives us some nice error reporting when there's a failure, too.

**Review by tgross — APPROVED** [review]
LGTM once the link gets fixed. Take my other comments as suggestions only.

Not sure why the tests didn't run on CircleCI. The new test code passes fine (including a run of all of `./command/agent`), and we'll probably get a full run once you push the fix for the link anyways:

```
$ go test -v ./command/agent -run TestConfig_normalizeAddrs
=== RUN   TestConfig_normalizeAddrs_DevMode
--- PASS: TestConfig_normalizeAddrs_DevMode (0.00s)
=== RUN   TestConfig_normalizeAddrs_NoAdvertise
--- PASS: TestConfig_normalizeAddrs_NoAdvertise (0.00s)
=== RUN   TestConfig_normalizeAddrs_AdvertiseLocalhost
--- PASS: TestConfig_normalizeAddrs_AdvertiseLocalhost (0.00s)
=== RUN   TestConfig_normalizeAddrs_IPv6Loopback
--- PASS: TestConfig_normalizeAddrs_IPv6Loopback (0.00s)
=== RUN   TestConfig_normalizeAddrs
--- PASS: TestConfig_normalizeAddrs (0.01s)
PASS
ok      github.com/hashicorp/nomad/command/agent        0.337s
```

File: website/content/docs/configuration/client.mdx [hunk 2]
```diff
@@ -48,7 +48,8 @@ client {
   to force network fingerprinting on. When run in dev mode, this defaults to the
   loopback interface. When not in dev mode, the interface attached to the
   default route is used. The scheduler chooses from these fingerprinted IP
-  addresses when allocating ports for tasks.
+  addresses when allocating ports for tasks. This value support [go-sockaddr/template
+  format][go-sockaddr/template].
```
[the remark this answers is not in the document]
Comment (angrycub) [reply website/content/docs/configuration/client.mdx:52]: thanks!

File: command/agent/config_test.go [hunk 3]
```diff
@@ -986,6 +987,103 @@ func TestConfig_normalizeAddrs(t *testing.T) {
 	}
 }
 
+func TestConfig_templateNetworkInterface(t *testing.T) {
```
Comment (tgross) [inline command/agent/config_test.go:990]: This is a nice testing improvement!

**Review by tgross — APPROVED** [review]

**Comment by shantanugadgil** [comment]
Could this be related to https://github.com/hashicorp/nomad/issues/8169 ?
Any chance of this landing in 1.1.0 ? :relaxed:

**Comment by tgross** [comment]
@shantanugadgil this changset will ship in 1.1.0. #8169 looks like a different field.

**Comment by shantanugadgil** [comment]
> looks like a different field.

Yes, I was making a push/request for the field `retry_join` for version `1.1.0` :grin:

**Pull request merged**